### PR TITLE
Cron support / Fix scientific notation / parallel sending

### PR DESCRIPTION
**Pull request by kernelogic**

### CODE REVIEW

File: src/replication/DealReplicationWorker.ts [hunk 1]
```diff
@@ -11,11 +11,13 @@ import GenerationRequest from '../common/model/GenerationRequest';
 const mathconfig = {};
 const math = create(all, mathconfig);
 const exec: any = require('await-exec');// no TS support
+const cron: any = require('node-cron');// no TS support
```
Comment (xinaxu) [inline src/replication/DealReplicationWorker.ts:14]: you can use https://www.npmjs.com/package/@types/node-cron

File: src/replication/DealReplicationWorker.ts [hunk 2]
```diff
@@ -55,10 +57,34 @@ export default class DealReplicationWorker extends BaseService {
...
     this.logger.debug(`${this.workerId} - Polling for work`);
+    await this.checkCronChange();
     const hasDoneWork = await this.pollReplicationWork();
     return hasDoneWork;
   }
 
+  private async checkCronChange (): Promise<void> {
+    const activeCronWorks = await Datastore.ReplicationRequestModel.find({
+      cronSchedule: { $ne: null },
+      workerId: { $ne: null },
```
Comment (xinaxu) [inline src/replication/DealReplicationWorker.ts:68]: this is checking cron works for all workers, but does it need to care about jobs for other worker? I think you can filter it to this worker Id.

Comment (kernelogic) [reply src/replication/DealReplicationWorker.ts:68]: There is only one worker at any time... so it doesn't matter I guess?

Comment (xinaxu) [reply src/replication/DealReplicationWorker.ts:68]: It needs to be scaled out in the future when possible.

File: src/replication/DealReplicationWorker.ts [hunk 3]
```diff
@@ -55,10 +58,34 @@ export default class DealReplicationWorker extends BaseService {
...
+      status: 'active'
+    });
+
+    for (let i = 0; i < activeCronWorks.length; i++) {
+      const request2Check = activeCronWorks[i];
+      if (this.cronRefArray.has(request2Check.id) && this.cronRefArray.get(request2Check.id)![0] !== request2Check.cronSchedule) {
```
Comment (xinaxu) [inline src/replication/DealReplicationWorker.ts:75]: Can you use https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements
So we get rid of the magic numbers

File: src/replication/DealReplicationWorker.ts [hunk 4]
```diff
@@ -16,6 +17,8 @@ export default class DealReplicationWorker extends BaseService {
   private readonly workerId: string;
   private readonly lotusCMD: string;
   private readonly boostCMD: string;
+  // holds reference to all started crons to be updated
+  private cronRefArray: Map<string, [string, ScheduledTask]> = new Map<string, [string, ScheduledTask]>();
```
Comment (liuziba) [inline src/replication/DealReplicationWorker.ts:21]: Try https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements
so we get rid of magic number 0 and 1

Comment (liuziba) [inline src/replication/DealReplicationWorker.ts:21]: Try https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements
so we get rid of magic number 0 and 1

**Review by liuziba — APPROVED** [review]

**Pull request merged**